schematic/base.py

```diff
@@ -0,0 +1,346 @@
+
+        Args:
+          value: The value to check, a string
+        Raises:
+          NotImplementedError: This should be implemented by subclasses.
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def get_parameter_for_value(value):
+        """Get the parameter for a parameterized implementation
+           of this class that is required to fit the given value.
+
+        Args:
+          value: the value to return the parameter for.
+        Raises:
+          NotImplementedError: Subclasses should implement this.
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def from_value(cls, value):
+        """Create an instance of this class that is compatible with value.
+
+        Args:
+          value: The value to return an instance for
+        Returns:
+          An instance of cls that can fit the value.
+        Raises:
+          ValueError: if value cannot fit in any instance of this class.
+        """
+        if cls.parameterized:
+            if not cls().is_value_compatible_with_class(value):
+                raise ValueError(
+                    "Value {} not compatible with any instance of {}".format(
+                        value, cls.name))
+            else:
+                return cls(parameter=cls.get_parameter_for_value(value))
+        else:
+            return cls()
+
+
+class TableColumn(DictableMixin, NameSqlMixin, object):
+    """DB-agnostic base class for storing info about a column in a table.
+
+    Attributes:
+      name: The identifier for the column
+      type: A TableColumnType instance with type information for the column.
+    """
+
+    def __init__(self, name, column_type):
+        self.name = name
+        self.type = column_type
+
+    def __repr__(self):
+        return "{}: {}".format(self.name, self.type)
+
+
+class TableDefinition(DictableMixin, NameSqlMixin, object):
+    """DB-agnostic base class for storing info about a table
+
+    Attributes:
+      name: The identifier for this table.
+      columns: A list of TableColumns describing this table's columns.
+    """
+
+    def __init__(self, name, columns):
+        self.name = name
+        self.columns = columns
+
+    def create_sql(self):
+        """Generate a sql statement for creating a table based on this TableDefinition.
+
+        Raises:
+          NotImplementedError: Subclasses should implement this.
+        """
+        raise NotImplementedError
+
+    def create_table(self, *args, **kwargs):
+        """Create the table in the destination
+            specified in *args and **kwargs
+
+        Raises:
+          NotImplementedError: Subclasses should implement this.
+        """
+        raise NotImplementedError
+
+    def add_column(self, column):
+        """Add a column to the column list.
+
+        Args:
+          column: A TableColumn instance to be added to this instance's columns.
+        """
+        for col in self.columns:
+            if col.name == column.name:
```

Comment:
```suggestion
            if col.name.upper() == column.name.upper():
```
Not relevant for every dbms, but maybe worth putting here to be safe